### PR TITLE
Feature 13861

### PR DESCRIPTION
**Pull request by mmoqui**
Add support for sensitive data. Only LDAP domains are for instance supported.
- First, the data that can be potential sensitive are declared in the descriptor of the user domain (an LDAP domain, that is a  domain whose users are defined into a remote LDAP service like AD or Azure AD)
- Second, in the administration page of the domain, the domain manager or the Silverpeas administrator can select the user for which the sensitivity of the data is applied.
Once the data privacy is applied to a user, those information aren't anymore returned to the requester, unless he's a Silverpeas administrator or a manager of the domain to which the user belong. This means the directories of users won't be able to render such sensitive information.

Don't forget to take into account also the PR: https://github.com/Silverpeas/Silverpeas-Components/pull/845

### CODE REVIEW

File: core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/arraypanes/AbstractArrayPane.java [hunk 1]
```diff
@@ -453,29 +379,18 @@ protected String printSortJavascriptFunction() {
...
     sb.append("helper: fixArrayPaneWidthHelper,");
-    if (StringUtil.isDefined(getUpdateSortJavascriptCallback())) {
-      sb.append("update: function(e, ui){").append(getUpdateSortJavascriptCallback()).append(";}");
-    }
     sb.append("}).disableSelection();");
     sb.append("</script>");
     return sb.toString();
   }
 
-  public String getUpdateSortJavascriptCallback() {
```
Comment (SilverYoCha) [inline core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/arraypanes/AbstractArrayPane.java:464]: I don't know about all signatures, but this one is used.
I think there is chance the others too...

File: core-web/src/main/java/org/silverpeas/core/webapi/profile/UserProfileEntity.java [hunk 2]
```diff
@@ -170,8 +170,8 @@ public String getLastName() {
 
   @Override
   @XmlElement
-  public String geteMail() {
-    return Encode.forHtml(this.user.geteMail());
+  public String getEmailAddress() {
```
Comment (SilverYoCha) [inline core-web/src/main/java/org/silverpeas/core/webapi/profile/UserProfileEntity.java:173]: Wow! I am very surprised that there is no JavaScript implementation impacts. Are you sure?

File: core-web/src/main/resources/META-INF/viewGenerator.tld [hunk 3]
```diff
@@ -1087,34 +1087,13 @@
...
-    <attribute>
-      <description>URL to be used to export data.</description>
-      <name>exportURL</name>
-      <required>false</required>
-      <rtexprvalue>false</rtexprvalue>
-      <type>java.lang.String</type>
-    </attribute>
-    <attribute>
-      <description>Defines if lines can be explicitly sorted</description>
-      <name>sortableLines</name>
```
Comment (SilverYoCha) [inline core-web/src/main/resources/META-INF/viewGenerator.tld:1099]: This is yet used (favorite links for example)

File: core-web/src/main/resources/META-INF/viewGenerator.tld [hunk 4]
```diff
@@ -1087,34 +1087,13 @@
...
     <attribute>
       <description>Defines number of lines displayed per page</description>
       <name>numberLinesPerPage</name>
       <required>false</required>
       <rtexprvalue>true</rtexprvalue>
       <type>java.lang.Integer</type>
     </attribute>
-    <attribute>
-      <description>Defines the javascript callback to call on a line move action</description>
-      <name>moveLineJsCallback</name>
```
Comment (SilverYoCha) [inline core-web/src/main/resources/META-INF/viewGenerator.tld:1113]: This is yet used (favorite links for example)

**Review by SilverYoCha — CHANGES_REQUESTED** [review]
I applaud the energy put into resolving SONAR feedback.
However, this has the consequence of adding a lot of fog to the reading of the code.
Indenting the code over the whole file instead of just the modified code does nothing to improve reading.
So I've given some feedback, but there may be some points missing.

Without having carried out any tests yet, I'm wondering how user notifications are sent, the user's email being a mandatory data item if the email channel is chosen.

File: core-library/src/main/java/org/silverpeas/core/admin/service/AdminController.java [hunk 5]
```diff
@@ -1527,11 +1359,20 @@ public List<UserDetail> getDeletedUsersInDomain(final String domainId) throws Ad
     return admin.getNonBlankedDeletedUsers(domainId);
   }
 
+  public List<UserDetail> getUsersWithSensitiveData(final String domainId) throws AdminException {
```
[the remark this answers is not in the document]
Comment (mmoqui) [reply core-library/src/main/java/org/silverpeas/core/admin/service/AdminController.java:1362]: Not all the `Administration`'s method invocations  are caught in `AdminController`. For example, the getting of removed or deleted users aren't caught in `AdminController`. In fact, my mind here is to follow the way the methods of Administration are invoked in the case of the functions related to the WWD menu (blocked users, deleted users, and so one).

File: core-library/src/main/java/org/silverpeas/core/admin/service/DefaultAdministration.java [hunk 6]
```diff
@@ -708,15 +694,6 @@ private void updateSpaceInheritance(SpaceInst space, boolean inheritanceBlocked)
     }
   }
 
-  @Override
```
[the remark this answers is not in the document]
Comment (mmoqui) [reply core-library/src/main/java/org/silverpeas/core/admin/service/DefaultAdministration.java:711]: Yes I'm sure

File: core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/arraypanes/ArrayPaneTag.java [hunk 7]
```diff
@@ -42,13 +41,9 @@ public class ArrayPaneTag extends TagSupport {
   private String var;
   private String title;
   private String summary;
-  private String isXHTML;
   private String routingAddress;
   private boolean export;
-  private String exportDataURL;
-  private boolean sortableLines;
```
[the remark this answers is not in the document]
Comment (mmoqui) [reply core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/arraypanes/ArrayPaneTag.java:49]: Shit. Yes indeed. They don't use the new sorting way (with column).  Only two JSP: `viewCategories.jsp` and `displayLinks.tag` are concerned

File: core-web/src/main/java/org/silverpeas/core/webapi/profile/UserProfileEntity.java (hunk 2 above)
Comment (mmoqui) [reply core-web/src/main/java/org/silverpeas/core/webapi/profile/UserProfileEntity.java:173]: Yes I'm sure now that I have just fixed some missing js.

File: core-web/src/main/resources/META-INF/viewGenerator.tld (hunk 4 above)
Comment (mmoqui) [reply core-web/src/main/resources/META-INF/viewGenerator.tld:1113]: Yes. Only in two jsp/tag. Shit

File: core-web/src/main/resources/META-INF/viewGenerator.tld (hunk 3 above)
Comment (mmoqui) [reply core-web/src/main/resources/META-INF/viewGenerator.tld:1099]: Yes only in two JSP/tag

File: core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/arraypanes/AbstractArrayPane.java (hunk 1 above)
Comment (mmoqui) [reply core-web/src/main/java/org/silverpeas/core/web/util/viewgenerator/html/arraypanes/AbstractArrayPane.java:464]: Huuu, it is used for move line js callback. Fuck

File: core-library/src/main/java/org/silverpeas/core/admin/service/DefaultOrganizationController.java [hunk 8]
```diff
@@ -1174,4 +1138,48 @@ public List<SpaceInstLight> getPathToComponent(String componentId) {
...
+        Stream.of(properties)
+            .map(userFull::getProperty)
+            .filter(DomainProperty::isSensitive)
+            .forEach(p -> userFull.setValue(p.getName(), null));
+      }
+    }
+    return user;
+  }
+
+  private boolean hasNotDomainAdminRights(User user) {
```
[the remark this answers is not in the document]
Comment (mmoqui) [reply core-library/src/main/java/org/silverpeas/core/admin/service/DefaultOrganizationController.java:1181]: Yes, indeed.

File: core-war/src/main/java/org/silverpeas/web/directory/model/UserItem.java [hunk 9]
```diff
@@ -114,27 +117,16 @@ public String getFax() {
...
-        .filter(UserDetail.class::isInstance)
-        .map(UserDetail.class::cast)
-        .orElseGet(() -> {
-          final UserDetail fromId = UserDetail.getById(getOriginalId());
-          if (fromId != null) {
-            userDetail = fromId;
-          }
-          return fromId;
-        });
+    return user instanceof UserDetail ? (UserDetail) user :
```
[the remark this answers is not in the document]
Comment (mmoqui) [reply core-war/src/main/java/org/silverpeas/web/directory/model/UserItem.java:120]: The cache is on UserFull. I just simplified the method. No need to update the wrapped user by another instance representing it.

File: core-war/src/main/webapp/jobDomainPeas/jsp/usersWithSensitiveData.jsp [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+<c:set var="language" value="${sessionScope.SilverSessionController.favoriteLanguage}"/>
+<fmt:setLocale value="${language}"/>
+<view:setBundle bundle="${requestScope.resources.multilangBundle}"/>
+
+<fmt:message var="usersWithSensitiveDataLabel" key="JDP.usersWithSensitiveData"/>
+
+<c:url var="displayUserWithSensitiveData" value="/RjobDomainPeas/jsp/displayUserWithSensitiveData"/>
+
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
+<html xmlns="http://www.w3.org/1999/xhtml" lang="${language}">
```
[the remark this answers is not in the document]
Comment (mmoqui) [reply core-war/src/main/webapp/jobDomainPeas/jsp/usersWithSensitiveData.jsp:41]: Ok

File: core-war/src/main/webapp/jobDomainPeas/jsp/userContent.jsp [hunk 11]
```diff
@@ -88,7 +91,7 @@
 
 <%@ include file="check.jsp" %>
 <%
-  Domain domObject = (Domain) request.getAttribute("domainObject");
+  //Domain domObject = (Domain) request.getAttribute("domainObject");
```
[the remark this answers is not in the document]
Comment (mmoqui) [reply core-war/src/main/webapp/jobDomainPeas/jsp/userContent.jsp:94]: Yes

File: core-library/src/main/java/org/silverpeas/core/admin/service/AdminController.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (SilverYoCha) [reply core-library/src/main/java/org/silverpeas/core/admin/service/AdminController.java:1362]: OK

File: core-war/src/main/java/org/silverpeas/web/directory/control/DirectoryUserFullRequestCache.java [hunk 12]
```diff
@@ -28,11 +28,7 @@
 import org.silverpeas.core.cache.service.CacheAccessorProvider;
 import org.silverpeas.web.directory.model.UserItem;
 
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
[the remark this answers is not in the document]
Comment (SilverYoCha) [reply core-war/src/main/java/org/silverpeas/web/directory/control/DirectoryUserFullRequestCache.java:31]: Mine is not doing that.

File: core-war/src/main/java/org/silverpeas/web/directory/model/UserItem.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (SilverYoCha) [reply core-war/src/main/java/org/silverpeas/web/directory/model/UserItem.java:120]: OK. No need to do more that you have done.

File: core-library/src/main/java/org/silverpeas/core/admin/service/DefaultOrganizationController.java [hunk 13]
```diff
@@ -1174,4 +1138,49 @@ public List<SpaceInstLight> getPathToComponent(String componentId) {
...
+    if (user != null && requester != null && !requester.getId().equals(user.getId()) &&
+        hasNotDomainAdminRights(requester, user.getDomainId())) {
+      return removeSensitiveDataForNonAdmin(user);
+    }
+    return user;
+  }
+
+  private <T extends UserDetail> T removeSensitiveDataForNonAdmin(T user) {
+    if (user.hasSensitiveData()) {
+      user.setEmailAddress("");
```
Comment (SilverYoCha) [inline core-library/src/main/java/org/silverpeas/core/admin/service/DefaultOrganizationController.java:1168]: Adding to my thoughts about my first reading of the PR, there is a problem with a user having sensitive data when it is targeted by a user notification when the targeted user has chosen email as a communication channel.

File: core-library/src/main/java/org/silverpeas/core/admin/service/DefaultOrganizationController.java [hunk 14]
```diff
@@ -1174,4 +1138,49 @@ public List<SpaceInstLight> getPathToComponent(String componentId) {
...
+  private UserDetail[] removeSensitiveData(final UserDetail[] userDetails) {
+    User requester = User.getCurrentRequester();
+    List<UserDetail> results = removeSensitiveData(List.of(userDetails), requester);
+    return results.toArray(new UserDetail[0]);
+  }
+
+  private <T extends UserDetail> List<T> removeSensitiveData(final List<T> users,
+      final User requester) {
+    if (!users.isEmpty() && requester != null) {
+      return users.stream()
```
Comment (SilverYoCha) [inline core-library/src/main/java/org/silverpeas/core/admin/service/DefaultOrganizationController.java:1151]: The given list could be a `SilverpeasList` one. So, the returned list MUST not be a new list instance. The list has just to be walked to apply on each User instance the sensitive data processing.

**Review by SilverYoCha — CHANGES_REQUESTED** [review]
Added new feedback

File: core-library/src/main/java/org/silverpeas/core/admin/service/cache/AdminCache.java [hunk 15]
```diff
@@ -261,10 +261,12 @@ private void removeUserDetail(String userId) {
 
   public Optional<UserDetail> getUserDetail(String userId) {
     if (useCache && useUserDetailCache) {
-      return Optional.ofNullable(userDetailCache.get(userId));
-    } else {
-      return Optional.empty();
+      UserDetail user = userDetailCache.get(userId);
```
[the remark this answers is not in the document]
Comment (mmoqui) [reply core-library/src/main/java/org/silverpeas/core/admin/service/cache/AdminCache.java:264]: After discussion with you, the decision was to use the copy constructor of UserDetail directly in DefaultAdministration#getUserDetail when returning the user details. The cache will kept the original user details.

File: core-library/src/main/java/org/silverpeas/core/admin/service/DefaultOrganizationController.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (mmoqui) [reply core-library/src/main/java/org/silverpeas/core/admin/service/DefaultOrganizationController.java:1168]: Yes it is done in my current fix (not yet pushed as I am resolving other issues)

Comment (mmoqui) [reply core-library/src/main/java/org/silverpeas/core/admin/service/DefaultOrganizationController.java:1168]: Yes it is done in my current fix (not yet pushed as I am resolving other issues)

File: core-library/src/main/java/org/silverpeas/core/admin/service/DefaultOrganizationController.java (hunk 14 above)
Comment (mmoqui) [reply core-library/src/main/java/org/silverpeas/core/admin/service/DefaultOrganizationController.java:1151]: Yes it is done in my current fix (not yet pushed as I am resolving other issues)

**Review by SilverYoCha — CHANGES_REQUESTED** [review]
Additional feedback

**Pull request merged**